vulkan/vulkan_device.cc

```diff
@@ -0,0 +1,320 @@
+    VkPhysicalDeviceFeatures* features) const {
+  if (features == nullptr || !physical_device_) {
+    return false;
+  }
+  vk.GetPhysicalDeviceFeatures(physical_device_, features);
+  return true;
+}
+
+bool VulkanDevice::GetPhysicalDeviceFeaturesSkia(uint32_t* sk_features) const {
+  if (sk_features == nullptr) {
+    return false;
+  }
+
+  VkPhysicalDeviceFeatures features = {0};
+  if (!GetPhysicalDeviceFeatures(&features)) {
+    return false;
+  }
+
+  uint32_t flags = 0;
+
+  if (features.geometryShader) {
+    flags |= kGeometryShader_GrVkFeatureFlag;
+  }
+  if (features.dualSrcBlend) {
+    flags |= kDualSrcBlend_GrVkFeatureFlag;
+  }
+  if (features.sampleRateShading) {
+    flags |= kSampleRateShading_GrVkFeatureFlag;
+  }
+
+  *sk_features = flags;
+  return true;
+}
+
+std::vector<VkQueueFamilyProperties> VulkanDevice::GetQueueFamilyProperties()
+    const {
+  uint32_t count = 0;
+
+  vk.GetPhysicalDeviceQueueFamilyProperties(physical_device_, &count, nullptr);
+
+  std::vector<VkQueueFamilyProperties> properties;
+  properties.resize(count, {0});
+
+  vk.GetPhysicalDeviceQueueFamilyProperties(physical_device_, &count,
+                                            properties.data());
+
+  return properties;
+}
+
+bool VulkanDevice::ChooseSurfaceFormat(const VulkanSurface& surface,
+                                       VkSurfaceFormatKHR* format) const {
+  if (!surface.IsValid() || format == nullptr) {
+    return false;
+  }
+
+  uint32_t format_count = 0;
+  if (VK_CALL_LOG_ERROR(vk.GetPhysicalDeviceSurfaceFormatsKHR(
+          physical_device_, surface.Handle(), &format_count, nullptr)) !=
+      VK_SUCCESS) {
+    return false;
+  }
+
+  if (format_count == 0) {
+    return false;
+  }
+
+  VkSurfaceFormatKHR formats[format_count];
+  if (VK_CALL_LOG_ERROR(vk.GetPhysicalDeviceSurfaceFormatsKHR(
+          physical_device_, surface.Handle(), &format_count, formats)) !=
+      VK_SUCCESS) {
+    return false;
+  }
+
+  for (uint32_t i = 0; i < format_count; i++) {
+    if (formats[i].format == VK_FORMAT_R8G8B8A8_UNORM) {
+      *format = formats[i];
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool VulkanDevice::ChoosePresentMode(const VulkanSurface& surface,
+                                     VkPresentModeKHR* present_mode) const {
+  if (!surface.IsValid() || present_mode == nullptr) {
+    return false;
+  }
+
+  uint32_t modes_count = 0;
+
+  if (VK_CALL_LOG_ERROR(vk.GetPhysicalDeviceSurfacePresentModesKHR(
+          physical_device_, surface.Handle(), &modes_count, nullptr)) !=
+      VK_SUCCESS) {
+    return false;
+  }
+
+  if (modes_count == 0) {
+    return false;
+  }
+
+  VkPresentModeKHR modes[modes_count];
+
+  if (VK_CALL_LOG_ERROR(vk.GetPhysicalDeviceSurfacePresentModesKHR(
+          physical_device_, surface.Handle(), &modes_count, modes)) !=
+      VK_SUCCESS) {
+    return false;
+  }
+
+  // Prefer the mailbox mode if available.
+  for (uint32_t i = 0; i < modes_count; i++) {
```

Comment:
Huh. Is there any case where I wouldn't use FIFO then? AFAIK, that is always available. I can get rid of this method entirely.